### PR TITLE
Support upload of analysis files

### PR DESCRIPTION
**Pull request by jakeale**
- Implement SQL model for `analysis_files`, a table to store analysis result files
- Support uploading at `POST /api/analyses/:id/files`
- Support downloading at `GET /api/analyses/:id/files/:id`
- Attach detailed file information from `analysis_files` SQL table to `GET /api/analyses/:id` calls

### CODE REVIEW

**Comment by igboyes** [comment]
I think the first three commits in this PR are from another branch.

**Comment by jakeale** [comment]
> I think the first three commits in this PR are from another branch.

Yes I think I messed up a rebase. It looks like they can be safely dropped if you don't mind another rebase.

File: tests/analyses/test_api.py [hunk 1]
```diff
@@ -235,6 +256,90 @@ async def test_remove(mocker, error, spawn_client, resp_is):
...
+        assert resp.status == 404
+
+    elif error == 422:
+        assert await resp_is.invalid_query(resp, {
+            "name": ["required field"]
+        })
+
+
+@pytest.mark.parametrize("exists", [True, False])
+async def test_download_file(exists, files, spawn_client, snapshot, tmpdir):
```
Comment (igboyes) [inline tests/analyses/test_api.py:314]: Can you figure out how to test that the correct file is actually written?

File: virtool/analyses/api.py [hunk 2]
```diff
@@ -174,6 +186,93 @@ async def remove(req: aiohttp.web.Request) -> aiohttp.web.Response:
...
+    if name_on_disk in files:
+        return conflict("File is already associated with analysis")
+
+    file_id = analysis_file["id"]
+    analysis_file_path = Path(req.app["settings"]["data_path"]) / "analyses" / name_on_disk
+
+    try:
+        size = await virtool.uploads.utils.naive_writer(req, analysis_file_path)
+    except asyncio.CancelledError:
+        logger.debug(f"Upload aborted: {file_id}")
```
Comment (igboyes) [inline virtool/analyses/api.py:230]: Change to "Analysis file upload" to differentiate from user uploads.

File: virtool/analyses/utils.py [hunk 3]
```diff
@@ -10,6 +14,25 @@
 )
 
 
+async def attach_analysis_files(pg: AsyncEngine, analysis_id: str) -> List[Dict]:
```
Comment (igboyes) [inline virtool/analyses/utils.py:17]: I don't think this is unit tested.

**Review by igboyes — CHANGES_REQUESTED** [review]

File: virtool/analyses/api.py [hunk 4]
```diff
@@ -174,6 +186,93 @@ async def remove(req: aiohttp.web.Request) -> aiohttp.web.Response:
...
+        return invalid_query(errors)
+
+    name = req.query.get("name")
+
+    if analysis_format not in ANALYSIS_FORMATS:
+        return bad_request("Unsupported analysis file format")
+
+    analysis_file = await virtool.analyses.db.create_row(pg, analysis_id, analysis_format, name)
+
+    name_on_disk = analysis_file["name_on_disk"]
```
Comment (igboyes) [inline virtool/analyses/api.py:218]: You need to refer to the files by their `id`, not `name_on_disk`.

File: virtool/analyses/api.py [hunk 5]
```diff
@@ -174,6 +186,93 @@ async def remove(req: aiohttp.web.Request) -> aiohttp.web.Response:
...
+        size = await virtool.uploads.utils.naive_writer(req, analysis_file_path)
+    except asyncio.CancelledError:
+        logger.debug(f"Analysis file upload aborted: {file_id}")
+        await virtool.analyses.db.delete_row(pg, file_id)
+
+        return aiohttp.web.Response(status=499)
+
+    analysis_file = await virtool.uploads.db.finalize(pg, size, file_id, AnalysisFile)
+
+    if not analysis_file:
```
Comment (igboyes) [inline virtool/analyses/api.py:237]: If this happens it is a bug. We should not handle it and blame it on the client (`4**` status).

Clients should also not receive messages about implementation details of the application such as problems with the backing SQL database. The API is meant to abstract away these details.

File: virtool/analyses/api.py [hunk 6]
```diff
@@ -174,6 +186,93 @@ async def remove(req: aiohttp.web.Request) -> aiohttp.web.Response:
...
+        await req.app["run_in_thread"](os.remove, analysis_file_path)
+        return not_found("Row not found in table after file upload")
+
+    files.append(name_on_disk)
+
+    await db.analyses.update_one({"_id": analysis_id}, {
+        "$set": {"files": files}
+    })
```
Comment (igboyes) [inline virtool/analyses/api.py:245]: Use `$push`. If there are two requests being handled at the same time (ie. the workflow is concurrently uploading two files), rewriting the `files` field could cause a serious bug.

File: virtool/analyses/db.py [hunk 7]
```diff
@@ -208,6 +214,71 @@ async def create(
     return document
 
 
+async def create_row(pg: AsyncEngine, analysis_id: str, analysis_format: str, name: str) -> Dict[str, any]:
```
Comment (igboyes) [inline virtool/analyses/db.py:217]: The naming of these is going to clash with the `Analysis` model when it is eventually created.

I think we should move the analysis file code into a new module: `virtool.analyses.files`. Also, change these function names to something like `create_analysis_file`.

File: virtool/analyses/db.py [hunk 8]
```diff
@@ -208,6 +214,71 @@ async def create(
...
+        await session.commit()
+
+
+async def get_row(pg: AsyncEngine, file_id: int) -> Optional[AnalysisFile]:
+    """
+    Get a row that represents an analysis result file by its `id`
+
+    :param pg: PostgreSQL AsyncEngine object
+    :param file_id: Row `id` to get
+    :return:
```
Comment (igboyes) [inline virtool/analyses/db.py:271]: Describe `:return:`.

File: virtool/analyses/models.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    tsv = "tsv"
+    json = "json"
+
+
+ANALYSIS_FORMATS = [*AnalysisFormat.to_list(), None]
+
+
+class AnalysisFile(Base):
+    """
+    SQL table to store new analysis files
```
Comment (igboyes) [inline virtool/analyses/models.py:31]: Is a model, not a table.

File: virtool/analyses/utils.py [hunk 10]
```diff
@@ -10,6 +14,25 @@
...
+    :param analysis_id: An id for a specific analysis
+    :return: List of file details for each file associated with an analysis
+    """
+    files = []
+
+    async with AsyncSession(pg) as session:
+        results = (await session.execute(select(AnalysisFile).filter_by(analysis=analysis_id))).scalars().all()
+
+        for result in results:
+            files.append(result.to_dict())
```
Comment (igboyes) [inline virtool/analyses/utils.py:31]:
```py
async with AsyncSession(pg) as session:
    results = (await session.execute(select(AnalysisFile).filter_by(analysis=analysis_id))).scalars().all()

return [result.to_dict() for result in results]
```

File: virtool/uploads/utils.py [hunk 11]
```diff
@@ -28,6 +28,11 @@ async def naive_writer(req, file_path) -> int:
 
     size = 0
 
+    try:
+        file_path.parent.mkdir()
```
Comment (igboyes) [inline virtool/uploads/utils.py:32]: Must run in thread to prevent filesystem blocking.

**Review by igboyes — CHANGES_REQUESTED** [review]

**Review by igboyes — APPROVED** [review]

**Pull request merged**